---
title:  "SCOP_2023_0255: WT and a1a2KO of AMPK in mouse hepatocytes"
author:
- "Analysis and data preparation for publication by Conchita Fraguas Bringas
based on prior data analysis by Christian Grønbæk & Lars Roed Ingerslev"
- "conchita.fraguas@sund.ku.dk"
output: html_document
date: '2024-06-03'
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## This notebook contains code for setting significance thresholds, calculating gene groups of interest, Venn diagrams, Heatmap figure composition, GO enrichment analysis and plot creation and Volcano plot creation alongside packages used.


##Let's load libraries first

```{r}
library(edgeR)
library(tidyverse)
library(limma)
library(ggrepel)
library(annotables)
library(pathview)
library(AnnotationDbi)
library(data.table)
library(here)
library(forcats)
library(purrr)
library(tibble)
library(S4Vectors)
library(plotly)
library(stringr)
library(readr)
library(Biobase)
library(limma)
library(lubridate)
library(IRanges)
library(BiocGenerics)
```

```{r}
library(tximport)
library(clusterProfiler)
library(DOSE)
library(org.Mm.eg.db)
library("biomaRt")
library("pheatmap")
library("graphics")
library("grDevices")
library("plotrix")
library("systemfonts")
library("ggplot2")
```

## prepare metadata sheet

```{r}
meta_data_sheet <- metadata_X1_SCOP_2023_0255
```

```{r}
dge_full_model <- DGE
dge_full_model$samples$group
```

```{r}
org_db <- org.Mm.eg.db

theme_set(theme_bw())
complex_setup <- TRUE
results_dir <- here("results")
dir.create(results_dir, showWarnings = F)
```

```{r}
formula_full_model <- formula("~0 + group + rin + cage_number + extraction_pool + pooled_status")
```

#data quality assessment

```{r}
n_samples <- nrow(dge_full_model$samples)
n_genes <-nrow(dge_full_model)
n_counts_per_sample <- colSums(dge_full_model$counts)
n_counts_total <-  sum(n_counts_per_sample)
n_counts_per_gene_per_sample <- (n_counts_total/n_genes)/n_samples
```

```{r}
fit <- `fit (2)`
```

```{r}
install.packages("devtools")
```

```{r}
devtools::install_github("https://github.com/CBMR-Single-Cell-Omics-Platform/cbmr")
```

```{r}
design <-  model.matrix(object = formula_full_model,
                        data = dge_full_model$samples)

colnames(design) <- str_remove_all(colnames(design), "group")
stopifnot(is.fullrank(design))
```

```{r}
technical_vars <- names(attr(design, "contrasts"))[-1]
technical_vars <- c("rin", technical_vars) |> set_names()

test_technical_vars <- function(var, fit) {
  coefs <- colnames(fit$design)[colnames(fit$design) %like% paste0("^", var)]
  cbmr::edgeR_tester(coef = coefs,
                     efit = fit,
                     id_name = "ENSEMBL")
}

deg_technical_vars <- map(technical_vars, test_technical_vars, fit = fit)
```

```{r}
contrast_matrix <- limma::makeContrasts(
  levels = design,

  #treatment effects
  wt_dmso_mk_vs_wt_dmso = WT_MK8722 - WT_DMSO, #1
  wt_bay_vs_wt_dmso = WT_BAY - WT_DMSO, #2
  wt_bay_mk_vs_wt_dmso = WT_BAY_MK8722 - WT_DMSO, #3
  wt_bay_vs_wt_mk = WT_BAY - WT_MK8722, #13
  wt_bay_mk_vs_wt_bay =  WT_BAY_MK8722 - WT_BAY, #15
  wt_bay_mk_vs_wt_dmso_mk =  WT_BAY_MK8722 - WT_MK8722, #16

  #.. and similarly for KO:
  ko_dmso_mk_vs_ko_dmso = KO_MK8722 - KO_DMSO, #4
  ko_bay_vs_ko_dmso = KO_BAY - KO_DMSO, #5
  ko_bay_mk_vs_ko_dmso = KO_BAY_MK8722 - KO_DMSO, #6
  ko_bay_vs_ko_mk = KO_BAY - KO_MK8722, #14
  ko_bay_mk_vs_ko_bay =  KO_BAY_MK8722 - KO_BAY, #16
  ko_bay_mk_vs_ko_mk = KO_BAY_MK8722 - KO_MK8722,

  #wt_bay_mk_vs_wt_mk = WT_BAY_MK8722 - WT_MK8722,
  #ko_bay_mk_vs_ko_mk = KO_BAY_MK8722 - KO_MK8722,

  #Cell type effects of bay?
  ko_bay_vs_wt_bay = KO_BAY - WT_BAY, #7
  #do controls work as expected? - effect of mk + dmso (dmso implicit part of mk treatment):
  ko_mk_vs_wt_mk = KO_MK8722 - WT_MK8722, #8
  #- effect of dmso
  ko_dmso_vs_wt_dmso = KO_DMSO - WT_DMSO, #9
  #Some change-cell_type-change-treatment cases:
  ko_dmso_vs_wt_bay = KO_DMSO - WT_BAY, #10
  ko_bay_vs_wt_bay_mk = KO_BAY - WT_BAY_MK8722, #11
  ko_mk_vs_wt_bay_mk = KO_MK8722 - WT_BAY_MK8722 #12
)

contrast_names <- colnames(contrast_matrix) |> set_names()
```

```{r}
deg_contrasts <- map(
  contrast_names,
  cbmr::edgeR_tester,
  contrast_matrix = contrast_matrix,
  efit = fit,
  id_name = "ENSEMBL")

invisible(lapply(deg_contrasts, `[`, , c("ENSEMBL", "F") := NULL))
deg_contrasts <- map(deg_contrasts, rename, avg_log2_CPM_project_wide=logCPM)
```

```{r}
deg_contrasts <- lapply(deg_contrasts, dplyr::rename,
                        Name = SYMBOL,
                        ENSEMBL_ID = ENSEMBL,
                        description = GENENAME)
```

```{r}
deg_contrasts$DMSO_vs_BAY_MK_interaction
```

```{r}
saveRDS(deg_contrasts,"/work/Home/deg_contrasts_new_BAY0255_Conchita_July_2024.RDS")
```

```{r}
install.packages("writexl")
```

```{r}
library(writexl)
```

```{r}
write_xlsx(deg_contrasts, "/work/Home/BAY_deg_contrasts_new_July_2024")
```
```{r}
write_xlsx(ontology_tests$Reactome, "/work/Home/0255_BAY/GO_Conchita_Aug_2024/GO_Reactome")
```

```{r}
get_GO_datatable_updated <- function(df, ...) {
  df %>%
    transmute(ID,
              Direction=as.factor(Direction),
              Ontology,
              TERM,
              `#genes`=NGenes,
              PValue=signif(PValue,2),
              FDR=signif(FDR, 2)) %>%
    DT::datatable(#extensions = 'Buttons',
      # filter="top",
      rownames = FALSE,
      options = list(dom = 'Bfrtip',
                     buttons = c('copy', 'csv', 'excel', 'colvis'),
                     pageLength = 10),
      height = 700,
      width = "100%",
      ...)
}
```

```{r}
#gene ontology
terms <-
  cbmr::get_enrichment_terms(
    org_db = org_db,
    gene_ids = rownames(fit),
    gene_id_key_type = "ENSEMBL",
    min_genes = 5,
    max_genes = 500,
    cache_path = "/work/Home/0255_BAY/shared_data/GO_terms")
```

```{r}
run_ontology_tests <- cbmr::run_ontology_tests(GO_terms,
                                           dge_full_model,
                                           contrast_matrix = contrast_matrix,
                                           fun=limma::camera)
```

```{r}
terms <-
  cbmr::get_enrichment_terms(
    org_db = org_db,
    gene_ids = rownames(fit),
    gene_id_key_type = "ENSEMBL",
    min_genes = 5,
    max_genes = 500,
    cache_path = "shared_data/GO_terms_cache")
```

```{r}
ontology_tests <- map(
  terms,
  cbmr::run_ontology_tests,
  y = dge_full_model,
  contrast_matrix = contrast_matrix,
  fun = limma::camera) |>
  suppressWarnings()
```

```{r}
merged_ontologies <- list()
for (i in names(ontology_tests)) {
  for (j in names(ontology_tests[[i]])) {
    this_table <- ontology_tests[[i]][[j]]
    this_table[["Ontology"]] <- i
    merged_ontologies[[j]] <- rbind(merged_ontologies[[j]], this_table)
  }
}

summary_table <- list(
  "Differentially expressed genes (FDR<5%)" = cbmr::n_sig_genes_pr_contrast(deg_contrasts),
  "Enriched GO-terms (FDR<5%)" = cbmr::n_sig_genes_pr_contrast(merged_ontologies)
  ) |>
  bind_cols() |>
  mutate(`Comparisons`=colnames(contrast_matrix)) |>
  relocate(`Comparisons`)
```

```{r}
summary_table
```

```{r}
get_GO_datatable_updated <- function(df, ...) {
  df %>%
    transmute(ID,
              Direction=as.factor(Direction),
              Ontology,
              TERM,
              `#genes`=NGenes,
              PValue=signif(PValue,2),
              FDR=signif(FDR, 2)) %>%
    DT::datatable(#extensions = 'Buttons',
      # filter="top",
      rownames = FALSE,
      options = list(dom = 'Bfrtip',
                     buttons = c('copy', 'csv', 'excel', 'colvis'),
                     pageLength = 10),
      height = 700,
      width = "100%",
      ...)
}
```

```{r}
for (contrast in contrast_names) {
  cbmr::catHeader_w_tabset(text = contrast, level = 2)

 cbmr::catHeader("Gene table (FDR<0.05)", level = 3)
  deg_contrasts[[contrast]] |>
    dplyr::filter(FDR<0.05) |>
    dplyr::slice_head(n = 100) |>
    cbmr::get_DE_datatable_updated() |>
    cbmr::print_DT()

  cbmr::catHeader("P-value distribution", level=3)
  deg_contrasts[[contrast]] |>
    pull("PValue") |>
    hist(main="P-value distribution of all genes tested") |>
    invisible()

  cbmr::catHeader("GO-term table (FDR<0.05)", level=3)
  merged_ontologies[[contrast]] |>
    filter(FDR<0.05) |>
    arrange(PValue) |>
    dplyr::slice_head(n = 100) |>
    get_GO_datatable_updated() |>
    cbmr::print_DT()
}
```

## Let's continue analysis into specific gene groups using the deg_constrasts object

```{r}
deg_contrasts <- deg_contrasts_new_BAY0255_Conchita_July_2024
```

```{r}
WT_BAY_MK_vs_WT_MK <- deg_contrasts$wt_bay_mk_vs_wt_dmso_mk

WT_BAY_vs_WT_DMSO <- deg_contrasts$wt_bay_vs_wt_dmso

WT_MK_vs_WT_DMSO <- deg_contrasts$wt_dmso_mk_vs_wt_dmso

WT_vs_KO_MK <- deg_contrasts$ko_mk_vs_wt_mk

WT_BAY_MK_vs_WT_DMSO <- deg_contrasts$wt_bay_mk_vs_wt_dmso
WT_BAY_vs_WT_MK <- deg_contrasts$wt_bay_vs_wt_mk

WT_vs_KO_DMSO <- deg_contrasts$ko_dmso_vs_wt_dmso

```

```{r}
#Set thresholds FC ≥ 1.3 (0.3785) and FDR < 0.05

sig_WT_vs_KO_DMSO <- WT_vs_KO_DMSO %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)
sig_WT_BAY_MK_vs_WT_MK <- WT_BAY_MK_vs_WT_MK %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)
sig_WT_BAY_vs_WT_DMSO <- WT_BAY_vs_WT_DMSO %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)
sig_WT_MK_vs_WT_DMSO <- WT_MK_vs_WT_DMSO %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)
sig_WT_vs_KO_MK <- WT_vs_KO_MK %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)
sig_WT_BAY_MK_vs_WT_DMSO <- WT_BAY_MK_vs_WT_DMSO %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)
sig_WT_BAY_vs_WT_MK <- WT_BAY_vs_WT_MK %>% filter(FDR < 0.05 & abs(logFC) >= 0.3785)

```

```{r}
#up and down thresholds

up_sig_WT_vs_KO_DMSO <- sig_WT_vs_KO_DMSO %>% filter(logFC>0.3785)
down_sig_WT_vs_KO_DMSO <- sig_WT_vs_KO_DMSO %>% filter(logFC< -0.3785)

up_sig_WT_BAY_MK_vs_WT_MK <- sig_WT_BAY_MK_vs_WT_MK %>% filter(logFC>0.3785)
down_sig_WT_BAY_MK_vs_WT_MK <- sig_WT_BAY_MK_vs_WT_MK %>% filter(logFC< -0.3785)

up_sig_WT_BAY_vs_WT_DMSO <- sig_WT_BAY_vs_WT_DMSO %>% filter(logFC>0.3785)
down_sig_WT_BAY_vs_WT_DMSO <- sig_WT_BAY_vs_WT_DMSO %>% filter(logFC< -0.3785)

up_sig_WT_MK_vs_WT_DMSO <- sig_WT_MK_vs_WT_DMSO %>% filter(logFC>0.3785)
down_sig_WT_MK_vs_WT_DMSO <- sig_WT_MK_vs_WT_DMSO %>% filter(logFC< -0.3785)

up_sig_WT_vs_KO_MK <- sig_WT_vs_KO_MK %>% filter(logFC>0.3785)
down_sig_WT_vs_KO_MK <- sig_WT_vs_KO_MK %>% filter(logFC< -0.3785)

down_sig_WT_BAY_MK_vs_WT_DMSO <- sig_WT_BAY_MK_vs_WT_DMSO %>% filter(logFC< -0.3785)

down_sig_WT_BAY_vs_WT_MK <- sig_WT_BAY_vs_WT_MK  %>% filter(logFC< -0.3785)

```


## All significant up/downregulated genes with FC ≥ 1.3; FDR < 0.05 thresholds

#1) MK-8722-stimulated-AMPK-dependent genes are those that are A) significantly donwregulated in AMPK DKO vs WT in MK-8722 and B) significantly upregulated in MK-8722 WT vs DMSO WT.

#2) BAY downregulated AMPK-stimulated genes are in the Venn diagram overlap between #1 and the significant downregulated genes in BAY + MK vs MK alone in WT.


```{r}
data <- up_sig_WT_MK_vs_WT_DMSO
data #1841 genes
```
```{r}
data2 <- down_sig_WT_vs_KO_MK
data2 #1973 genes
```
```{r}
AMPK_genes <- data %>% dplyr::filter(Name %in% (data2$Name))
AMPK_genes #965 genes
```

```{r}
summary(duplicated(AMPK_genes$Name))
```

```{r}
duplicate_rows <- duplicated(AMPK_genes$Name)
AMPK <- AMPK_genes[!duplicate_rows, ]
summary(AMPK) #846 AMPK genes
```
```{r}
duplicate_rows <- duplicated(AMPK$ENSEMBL_ID)
AMPK2 <- AMPK[!duplicate_rows, ]
summary(AMPK2) #843 AMPK genes
```
```{r}
AMPK3 <- na.omit(AMPK2)
AMPK3 #842
```
```{r}
selected <- logCPM %>% dplyr::filter(SYMBOL %in% (AMPK_genes$Name))
selected #844 downreg BAY MK genes :) in logCPM data
```
```{r}
duplicate_rows <- duplicated(selected$SYMBOL)
clean_data <- selected[!duplicate_rows, ]
summary(clean_data) #845 AMPK genes in logCPM dataset good
```
```{r}
final <- na.omit(clean_data)
final #845 genes
symbol <- final$SYMBOL
```
```{r}
selected2  <- final %>% dplyr::select(starts_with("X0255"))
selected2_df <- selected2 %>% as.data.frame()
selected2_df
```

```{r}
rownames(selected2_df) = symbol
selected2_df #845 MK-AMPK genes
```
```{r}
## plot a heatmap of MK-stimulated-AMPK-dependent genes
pheatmap(selected2_df, annotation = Conch_edits_metadata_good_final_0255_ %>% column_to_rownames("sample_id") %>% dplyr::select("treatment"),
           col= colors.Conch, border_color=NA, fontsize = 10,
         fontsize_row = 10, height=20, show_rownames = FALSE, width = 20, scale = "row")
```
```{r}
write.csv (final, "/work/Home/MK_AMPK_genes_845_in_logCPM_per_sample.csv")
```

```{r}
BAY_down_AMPK_genes2 <- final %>% dplyr::filter(SYMBOL %in% (down_sig_WT_BAY_MK_vs_WT_MK$Name))
BAY_down_AMPK_genes2 #524 genes
```
```{r}
write.csv(BAY_down_AMPK_genes2 , "/work/Home/BAY_downreg_AMPK_genes_524_in_logCPM_per_sample.csv")
```
## Preparing data for heatmap visualisation

```{r}
down_sig_WT_BAY_MK_vs_WT_MK
```
```{r}
selected <- logCPM %>% dplyr::filter(SYMBOL %in% (BAY_down_AMPK_genes$SYMBOL))
selected #525 downreg BAY MK genes :) in logCPM data
```
```{r}
duplicate_rows <- duplicated(selected$SYMBOL)
clean_data <- selected[!duplicate_rows, ]
summary(clean_data) #524 AMPK genes in logCPM dataset good
# 524 of MK-stimulated-AMPK-dependent genes are downregulated by BAY-3827 when AMPK is active (+ MK-8722)
```
```{r}
final <- na.omit(selected)
final #845 genes
symbol <- final$SYMBOL
```
```{r}
selected2  <- final %>% dplyr::select(starts_with("X0255"))
selected2_df <- selected2 %>% as.data.frame()
selected2_df
```

```{r}
#rownames(selected2_df) = symbol
selected2_df #845 MK-AMPK genes
```
```{r}
pheatmap(selected2_df, annotation = Conch_edits_metadata_good_final_0255_ %>% column_to_rownames("sample_id") %>% dplyr::select("treatment"),
           col= colors.Conch, border_color=NA, fontsize = 10,
         fontsize_row = 10, height=20, show_rownames = FALSE, width = 20, scale = "row")
```

## Venn diagrams
```{r}

install.packages("ggvenn")
```

```{r}
library(ggvenn)
```

```{r}
listA1 <-  as.list(final$SYMBOL)
listB2 <- as.list(down_sig_WT_BAY_MK_vs_WT_MK$Name)
listA1
```

```{r}
#ALL upregulated genes figure MK and Torin vs DMSO in WT and KO
genes <- paste("gene",1:1000,sep="")
x <- list(
 MK_AMPK_genes_845 = listA1,
 down_BAY_MK_vs_WT_MK = listB2


  )
ggvenn(x
  ,
  fill_color = c("#0073C2FF","#868686FF"),
  stroke_size = 0.5, set_name_size = 5
  )
```
## Clusterprofiler GO enrichment analysis FDR < 0.05

```{r}
## Create background dataset for hypergeometric testing using all genes tested for significance in the results
#GO for WT_vs_KO in MK-8722 treatment and selecting for downregulated significant genes

allOE_genes <- dplyr::filter(WT_vs_KO_MK, !is.na(ENSEMBL_ID)) %>%
  pull(ENSEMBL_ID) %>%
  as.character()
allOE_genes
```
```{r}
## Extract significant results
sigOE <- dplyr::filter(down_sig_WT_vs_KO_MK, FDR < 0.05 & !is.na(ENSEMBL_ID))

sigOE_genes <- sigOE %>%
  pull(ENSEMBL_ID) %>%
  as.character()
sigOE_genes
```
```{r}
# AND GO for WT_BAY_MK_vs_WT_MK treatment and selecting for downregulated significant genes
allOE_genes <- dplyr::filter(WT_BAY_MK_vs_WT_MK, !is.na(ENSEMBL_ID)) %>%
  pull(ENSEMBL_ID) %>%
  as.character()
allOE_genes
```
```{r}
## Extract significant results
sigOE <- dplyr::filter(down_sig_WT_BAY_MK_vs_WT_MK, FDR < 0.05 & !is.na(ENSEMBL_ID))

sigOE_genes <- sigOE %>%
  pull(ENSEMBL_ID) %>%
  as.character()
sigOE_genes
```

```{r}
## Run GO enrichment analysis
ego <- enrichGO(gene = sigOE_genes,
                universe = allOE_genes,
                keyType = "ENSEMBL",
                OrgDb = org.Mm.eg.db,
                ont = "ALL",
                pAdjustMethod = "BH",
                qvalueCutoff = 0.05,
                readable = TRUE)
```

```{r}
#Let's check the results:
## Output results from GO analysis to a table
cluster_summary <- data.frame(ego)
cluster_summary
```

```{r enrichGO_dotplot, fig.height= 10, fig.width=5}
## Dotplot
dotplot(ego, showCategory=50,font.size = 6)
```

```{r}
# Set fold changes
## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector
OE_foldchanges <- sigOE$Name
names(OE_foldchanges) <- sigOE$Name


```{r}
## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)

cnetplot(ego,
         categorySize="FDR",
         showCategory = 5,
         foldChange=OE_foldchanges,
         #layout = "star",
         vertex.label.font=4, cex_label_gene = 0.5)
```

## Volcano plots FC => 1.3, FDR < 0.05
#Based on tutorial by https://biostatsquid.com/volcano-plots-r-tutorial/
```{r}
## Obtain logical vector where TRUE values denote padj values < 0.05 and fold change > 1.3 in either direction
WT_vs_KO_MK_tb <- WT_vs_KO_MK %>%
  as_tibble(rownames = "gene")
```

```{r}
WT_vs_KO_MK_tb <- WT_vs_KO_MK_tb %>%
mutate(threshold = FDR < 0.05 & abs(logFC) >= 0.3785)
```

```{r volcano_plot}
## Volcano plot
ggplot(WT_vs_KO_MK_tb) +
  geom_point(aes(x = logFC, y = -log10(FDR), colour = threshold)) +
  ggtitle("WT vs AMPK DKO in MK-8722") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  scale_y_continuous(limits = c(0,50)) +
  theme(legend.position = "none",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25)))  
```

```{r}
## Create an empty column to indicate which genes to label
WT_vs_KO_MK_tb <- WT_vs_KO_MK_tb %>% mutate(genelabels = "")
```

```{r}
## Sort by padj values
WT_vs_KO_MK_tb <- WT_vs_KO_MK_tb %>% arrange(FDR)
```

```{r}
## Populate the gene labels column with contents of the gene symbols column for the first 10 rows, i.e. the top 10 most significantly expressed genes
WT_vs_KO_MK_tb$genelabels[1:30] <- as.character(WT_vs_KO_MK_tb$SYMBOL[1:30])

head(WT_vs_KO_MK_tb)
```

```{r volcano_plot_label}
ggplot(WT_vs_KO_MK_tb, aes(x = logFC, y = -log10(FDR))) +
  geom_point(aes(colour = threshold)) +
  geom_text_repel(aes(label = genelabels)) +
  ggtitle("WT vs AMPK DKO in MK-8722") +
  xlab("log2 fold change") +
  ylab("-log10 adjusted p-value") +
  theme(legend.position = "right",
        plot.title = element_text(size = rel(1.5), hjust = 0.5),
        axis.title = element_text(size = rel(1.25))) + geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") +
  geom_vline(xintercept = c(log2(0.3785), log2(2)),
             linetype = "dashed")
```

```{r}
# Create new categorical column ------------------------------------------------
WT_vs_KO_MK_tb <- WT_vs_KO_MK_tb %>%
  mutate(gene_type = case_when(logFC >= 0.3785 & FDR <= 0.05 ~ "up",
                               logFC <= 0.3785 & FDR <= 0.05 ~ "down",
                               TRUE ~ "ns"))  
```

```{r}
WT_vs_KO_MK_tb %>%
  distinct(gene_type) %>%
  pull()  
```
```{r}
# Add colour, size and alpha (transparency) to volcano plot --------------------
cols <- c("up" = "red", "down" = "#26b3ff", "ns" = "grey")
sizes <- c("up" = 1, "down" = 1, "ns" = 1)
alphas <- c("up" = 1, "down" = 1, "ns" = 0.5)
```


```{r}
WT_vs_KO_MK_tb %>%
  ggplot(aes(x = logFC, y = -log10(FDR),
             fill = gene_type,    
             size = gene_type,
             alpha = gene_type)) +
  geom_point(shape = 21, # Specify shape and colour as fixed local parameters    
             colour = "black") +
  geom_hline(yintercept = -log10(0.05),
             linetype = "dashed") +
  geom_vline(xintercept = c(log2(0.5), log2(2)),
             linetype = "dashed") +
  scale_fill_manual(values = cols) + # Modify point colour
  scale_size_manual(values = sizes) + # Modify point size
  scale_alpha_manual(values = alphas) + # Modify point transparency
  scale_x_continuous(breaks = c(seq(-10, 10, 2)),       
                     limits = c(-6, 6))+
geom_text_repel(aes(label = genelabels), size = 1) +
  ggtitle("WT vs AMPK DKO in MK-8722") +
  xlab("Log2 Fold Change (log2FC)") +
  ylab("-Log10 False Discovery Rate (FDR)") +
  theme(legend.position = "right")
```
